include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -244,144 +179,267 @@ typedef enum GraphOptimizationLevel {
-
-ORT_API_STATUS(OrtRunOptionsSetRunLogVerbosityLevel, _Inout_ OrtRunOptions* options, int value);
-ORT_API_STATUS(OrtRunOptionsSetRunLogSeverityLevel, _Inout_ OrtRunOptions* options, int value);
-ORT_API_STATUS(OrtRunOptionsSetRunTag, _In_ OrtRunOptions*, _In_ const char* run_tag);
-
-ORT_API_STATUS(OrtRunOptionsGetRunLogVerbosityLevel, _In_ const OrtRunOptions* options, _Out_ int* out);
-ORT_API_STATUS(OrtRunOptionsGetRunLogSeverityLevel, _In_ const OrtRunOptions* options, _Out_ int* out);
-ORT_API_STATUS(OrtRunOptionsGetRunTag, _In_ const OrtRunOptions*, _Out_ const char** out);
+struct OrtKernelInfo;
+typedef struct OrtKernelInfo OrtKernelInfo;
+struct OrtKernelContext;
+typedef struct OrtKernelContext OrtKernelContext;
+struct OrtCustomOp;
+typedef struct OrtCustomOp OrtCustomOp;
 
-// Set a flag so that ALL incomplete OrtRun calls that are using this instance of OrtRunOptions
-// will exit as soon as possible.
-ORT_API_STATUS(OrtRunOptionsSetTerminate, _Inout_ OrtRunOptions* options);
-// Unset the terminate flag to enable this OrtRunOptions instance being used in new OrtRun calls.
-ORT_API_STATUS(OrtRunOptionsUnsetTerminate, _Inout_ OrtRunOptions* options);
+typedef enum OrtAllocatorType {
+  OrtDeviceAllocator = 0,
+  OrtArenaAllocator = 1
+} OrtAllocatorType;
 
 /**
- * Create a tensor from an allocator. OrtReleaseValue will also release the buffer inside the output value
- * \param out Should be freed by calling OrtReleaseValue
- * \param type must be one of TENSOR_ELEMENT_DATA_TYPE_xxxx
- */
-ORT_API_STATUS(OrtCreateTensorAsOrtValue, _Inout_ OrtAllocator* allocator,
-               _In_ const int64_t* shape, size_t shape_len, ONNXTensorElementDataType type,
-               _Outptr_ OrtValue** out);
+ * memory types for allocator, exec provider specific types should be extended in each provider
+ * Whenever this struct is updated, please also update the MakeKey function in onnxruntime/core/framework/execution_provider.cc
+*/
+typedef enum OrtMemType {
+  OrtMemTypeCPUInput = -2,              // Any CPU memory used by non-CPU execution provider
+  OrtMemTypeCPUOutput = -1,             // CPU accessible memory outputted by non-CPU execution provider, i.e. CUDA_PINNED
+  OrtMemTypeCPU = OrtMemTypeCPUOutput,  // temporary CPU accessible memory allocated by non-CPU execution provider, i.e. CUDA_PINNED
+  OrtMemTypeDefault = 0,                // the default allocator for execution provider
+} OrtMemType;
 
-/**
- * Create a tensor with user's buffer. You can fill the buffer either before calling this function or after.
- * p_data is owned by caller. OrtReleaseValue won't release p_data.
- * \param out Should be freed by calling OrtReleaseValue
- */
-ORT_API_STATUS(OrtCreateTensorWithDataAsOrtValue, _In_ const OrtMemoryInfo* info,
-               _Inout_ void* p_data, size_t p_data_len, _In_ const int64_t* shape, size_t shape_len,
-               ONNXTensorElementDataType type, _Outptr_ OrtValue** out);
+struct OrtApi {
```

Comment:
>OrtApi [](start = 7, length = 6)

Where do we add new interfaces? Is it always at the end? Logically they may group with some existing ones.
Also, have you considered exposing Session as a separate API?

Reply:
Yeah, I did start to think of dividing up the interface into separate ones but that would be a next step. Each change requires modifying all the client code that refers to it...